### PR TITLE
HPCC-20463 Report thor job queue from ThorQueueMonitor

### PR DESCRIPTION
**Pull request by wangkx**
A new WsWorkunits method WUGetThorJobQueue is added to report
the information about thor job queue from ThorQueueMonitor.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead  please review.

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp [hunk 1]
```diff
@@ -1784,3 +1786,70 @@ bool CWsWorkunitsEx::onWUClusterJobSummaryXLS(IEspContext &context, IEspWUCluste
...
+
+        const char *cluster = req.getCluster();
+        const char *startDate = req.getStartDate();
+        const char *endDate = req.getEndDate();
+        unsigned maxJobQueueItemsToReturn = req.getMaxJobQueueItemsToReturn();
+        if (maxJobQueueItemsToReturn == 0)
+            maxJobQueueItemsToReturn = DefaultMaxJobQueueItemsInWUGetJobQueueResponse;
+
+        CDateTime fromTime, toTime;
+        StringBuffer fromstr, tostr;
```
Comment (RussWhitehead) [inline esp/services/ws_workunits/ws_workunitsAuditLogs.cpp]: Should the time be set to 0:0:0 for fromstr, and 23:59:59 for tostr?  Otherwise it will only include part of the specified date

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp [hunk 2]
```diff
@@ -1784,3 +1786,70 @@ bool CWsWorkunitsEx::onWUClusterJobSummaryXLS(IEspContext &context, IEspWUCluste
...
+                continue;
+
+            jq.getAuditLineInfo(line, longestQueue, maxConnected, maxJobQueueItemsToReturn,
+                idx < lastLineID ? 1 : 2, queues);
+
+            if (queues.length() > maxJobQueueItemsToReturn)
+                break;
+        }
+
+        if (queues.length() > maxJobQueueItemsToReturn)
```
Comment (RussWhitehead) [inline esp/services/ws_workunits/ws_workunitsAuditLogs.cpp]: Could this code be moved to the conditional in line 1834?

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp [hunk 3]
```diff
@@ -1784,3 +1786,70 @@ bool CWsWorkunitsEx::onWUClusterJobSummaryXLS(IEspContext &context, IEspWUCluste
...
+        {
+            queues.pop();
+            VStringBuffer msg("More than %u job queues are found. Only %u job queues are returned.",
```
Comment (RussWhitehead) [inline esp/services/ws_workunits/ws_workunitsAuditLogs.cpp]: Would there be any benefit in reporting how many job queues were found (queues.length())?

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp [hunk 4]
```diff
@@ -2974,7 +2974,7 @@ bool WsWuJobQueueAuditInfo::checkNewThorQueueItem(IEspThorQueue* tq, unsigned sh
             bAdd = true;
         if (!checkSameStrings(tq->getConnectedThors(), tq0.getConnectedThors()))
             bAdd = true;
-        if (!checkSameStrings(tq->getConnectedThors(), tq0.getConnectedThors()))
+        if (!checkSameStrings(tq->getWaitingThors(), tq0.getWaitingThors()))
```
Comment (RussWhitehead) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp]: What is the purpose of this change?

**Review by RussWhitehead — COMMENTED** [review]
@wangkx A few questions

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp]: Good point. I will check and, if needed, append 0:0:0 or 23:59:59.

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp (hunk 2 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp]: Will do.

File: esp/services/ws_workunits/ws_workunitsAuditLogs.cpp (hunk 3 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsAuditLogs.cpp]: I avoid that just in case there are too many queues to process.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp (hunk 4 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp]: The old code is a duplicate of line 2975/2976. I guess that original author did a copy, but forgot to change it to getWaitingThors().

**Comment by wangkx** [comment]
I am closing this PR because more functions have to be added.
